File: common/src/main/java/dev/obscuria/fragmentum/v2/api/common/signal/Signal2.java
package dev.obscuria.fragmentum.v2.api.common.signal;

import dev.obscuria.fragmentum.v2.core.common.signal._Signal2;

@SuppressWarnings("unused")
public interface Signal2<P1, P2> extends Signal<Signal2.Listener<P1, P2>> {

    static <P1, P2> Signal2<P1, P2> create() {
        return new _Signal2<>();
    }

    void emit(P1 p1, P2 p2);

    @FunctionalInterface
    interface Listener<P1, P2> {

        void consume(P1 p1, P2 p2);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/api/tools/signal/Signal2.kt
package dev.obscuria.fragmentum.api.tools.signal

interface Signal2<P1, P2> : Signal<Signal2.Listener<P1, P2>>
{
    fun emit(p1: P1, p2: P2)

    fun interface Listener<P1, P2>
    {
        fun consume(p1: P1, p2: P2)
    }
}